Add all node addition and deletion functions

LeanderGlanda opened the pull request:
This pull request adds most of the functions for adding and deleting nodes, listed here: https://www.open62541.org/doc/master/server.html#node-addition-and-deletion

Not all functions have been implemented yet, I will add them too, I think. So take this as WIP.

Is it okay how I implemented the functions or should this be done in a different way?

Should these be tested or should examples be provided?

LeanderGlanda commented:
I changed a lot of things.
Now there is only a single addNode function, which should be able to add all different kinds of nodes to the server. This is possible because of features Rust has and C doesn't. And I guess from a usability point of view this is better.

Note that this doesn't work yet, something with the attributes causes a segfault, which I'm currently figuring out why it happens, I already have an idea.

LeanderGlanda commented:
What do you think, should the wrapper make using the server more convenient than using the C api or should it just crate a safe barrier between Rust and C?
@sgoll

LeanderGlanda commented on src/ua/data_types.rs:
Line endings were configured in a wrong way in VS code apparently. I fixed it now and also edited settings to apply cargo fmt on save.

LeanderGlanda commented:
I just fixed the formatting issues. As said in the issue thread, my VS code configuration wasn't set up properly.

I'll take a closer look at the builder methods idea now, it definitely sounds worth investigating. 

Regarding your answer to my question: Sounds like a good convention for this crate. This means I will keep up that all the different ObjectTypeAttributes/.*Attributes types are available as is for now. I had an idea to put them in an enum, and only in an enum, but that brings lots of different problems with it.
Right now every data_type has its own file. But most of the Attribute types are only 9 lines long. Should we keep it like this or move them together in a single file?

uklotzde commented on .vscode/settings.json:
```diff
@@ -0,0 +1,5 @@
+{
```
These personal settings of you IDE do not belong in the project repo.

Please remove this file. And maybe also add an exclusion rule to your global .gitignore file. We prefer not to add rules for custom tools to the project's .gitignore.

@sgoll If you agree, of course ;)

LeanderGlanda replied:
Okay, I was unsure if I should add them or not. I will just keep them on my machine.

uklotzde commented:
It is also recommended to run the pre-commit checks locally. `cargo install just` and `just setup` will get you started. Use `just pre-commit` for running the checks.

Alternatively run the commands from the `.justfile` manually.

LeanderGlanda commented:
I started implementing an example from open62541 (https://www.open62541.org/doc/master/tutorial_server_object.html) using this crate. This uncovers lots of things which are missing and need work.
I wanted to push my current state here, pre-commit checks haven't been run as I don't have more time today. Everything will be fixed after getting the example to work.

I've added a new method to the data_type trait: `.as_raw()`
Its use is to access the inner type directly without giving up ownership of the memory. This is sometimes required, as some C functions just take things by value and either just copy the content or only use it for the scope of the function, without caring about freeing memory.
It has to be used with care, that's why I added the unsafe keyword, but when the use is appropriate it makes things way better to use.
What do you think of that new method @sgoll ? It it fine or does it have a fundamental flaw which I didn't think of? Is there any other nice way to handle the "C pass-by-value without thinking about ownership which Rust doesn't have" issue?
Don't stress with the answer, it doesn't block my work currently.

LeanderGlanda commented:
I will probably split the changes into either pull requests or different commits when I'm done with the node addition and deletion.

LeanderGlanda commented:
Finally the object_type tutorial/example from open62541 documentation is implemented using this crate. It involved a lot of changes on top of just node addition and deletion.

For now I'll continue getting everything ready for merging into main. I'll open some pull requests for that or comment changes on the existing ones, if necessary.

After that I'll probably work on implementing the variable_type tutorial/example from open62541 docs.

Regarding the changes to adding nodes:
What do you think about how the node adding works?
How could it be done better? Right now I don't think it is very nice.

LeanderGlanda commented:
I've managed to split apart a lot of changes from here into single commits, see the following branches latest commits for that:

https://github.com/LeanderGlanda/open62541-crate/tree/implement_server_object_example_clean_commits


(This information may be completely useless)

LeanderGlanda commented:
Everything expect the node type lifecycle and the example are already merged. These two will not for the time being.

If you are from the future and happen to need the lifecycle feature from #145, talk to the maintainers / sgoll about the requirements and then find a solution together, as #145 was not ready to be merged.
